### DISCUSSION TITLE
[gatsby-core-utils] Retry logic on error for fetchRemoteFile

### DISCUSSION THREAD

**Discussion by njbmartin** [gatsbyjs/gatsby#33857] [Ideas / Feature Requests]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

Getting the following error when downloading image assets from Contentful (using the source plugin):

`RequestError: Client network socket disconnected before secure TLS connection was established`

When I see this message, it's an instant failure with no retry logic.

I noticed that the old "download retry logic" was replaced with Gatsby's `createRemoteFileNode`, so dug all the way down to `fetchRemoteFile` within `gatsby-core-utils`, and noticed there was `STALL_RETRY_LIMIT` and `INCOMPLETE_RETRY_LIMIT`, but no `ERROR_RETRY_LIMIT`.

When the responseStream errors, there is no retry logic:

https://github.com/gatsbyjs/gatsby/blob/aa050d7bb7857a98a4b3b9be92058bff8c4400f0/packages/gatsby-core-utils/src/fetch-remote-file.ts#L354-L365

The workaround was to add retry logic to this:

```js
    responseStream.on(`error`, error => {
      if (timeout) {
        clearTimeout(timeout);
      }

      fsWriteStream.close();

      fs.removeSync(tmpFilename);

      process.nextTick(() => {
        if (attempt < STALL_RETRY_LIMIT) {
          // Retry by calling ourself recursively
          resolve(requestRemoteNode(url, headers, tmpFilename, httpOptions, attempt + 1));
        } else {
          reject(error);
        }
      });
    });
```

As for why I'm getting `Client network socket disconnected before secure TLS connection was established`, it's probably something to do with me working inside Docker containers, so not flagging that as a Gatsby issue.

### Reproduction Link

N/A

### Steps to Reproduce

N/A (hard to reproduce)

### Expected Result

Any connection errors in `fetchRemoteFile` should at least be retried.

### Actual Result

Instant build failure

### Environment

```shell
System:
    OS: Linux 5.10 Debian GNU/Linux 10 (buster) 10 (buster)
    CPU: (12) x64 AMD Ryzen 5 3600 6-Core Processor
    Shell: 5.0.3 - /bin/bash
  Binaries:
    Node: 14.17.6 - /usr/local/bin/node
    Yarn: 1.22.5 - /usr/bin/yarn
    npm: 6.14.15 - /usr/local/bin/npm
  Languages:
    Python: 2.7.16 - /usr/bin/python
  npmPackages:
    gatsby: ^4.1.0 => 4.1.0
    gatsby-cli: ^4.0.0 => 4.1.0
    gatsby-plugin-canonical-urls: ^4.0.0 => 4.1.0
    gatsby-plugin-create-client-paths: ^4.0.0 => 4.1.0
    gatsby-plugin-env-variables: ^2.0.0 => 2.1.0
    gatsby-plugin-gatsby-cloud: ^4.0.0 => 4.1.0
    gatsby-plugin-google-tagmanager: ^4.0.0 => 4.1.0
    gatsby-plugin-image: ^2.0.0 => 2.1.0
    gatsby-plugin-loadable-components-ssr: ^3.1.0 => 3.4.0
    gatsby-plugin-manifest: ^4.0.0 => 4.1.0
    gatsby-plugin-nprogress: ^4.0.0 => 4.1.0
    gatsby-plugin-offline: ^5.0.0 => 5.1.0
    gatsby-plugin-perf-budgets: ^0.0.18 => 0.0.18
    gatsby-plugin-preconnect: ^1.1.54 => 1.3.0
    gatsby-plugin-preload-fonts: ^3.0.0 => 3.1.0
    gatsby-plugin-purgecss: ^6.1.0 => 6.1.0
    gatsby-plugin-react-helmet: ^5.0.0 => 5.1.0
    gatsby-plugin-remove-serviceworker: ^1.0.0 => 1.0.0
    gatsby-plugin-s3: ^0.3.8 => 0.3.8
    gatsby-plugin-schema-snapshot: ^3.0.0 => 3.1.0
    gatsby-plugin-sharp: ^4.0.0 => 4.1.0
    gatsby-plugin-sitemap: ^5.0.0 => 5.1.0
    gatsby-plugin-theme-ui: ^0.12.0 => 0.12.0
    gatsby-plugin-typescript: ^4.0.0 => 4.1.0
    gatsby-plugin-webpack-bundle-analyser-v2: ^1.1.25 => 1.1.25
    gatsby-source-contentful: ^6.0.0 => 6.1.0
    gatsby-source-filesystem: ^4.0.0 => 4.1.0
    gatsby-source-shopify: ^4.0.0 => 4.7.1
    gatsby-transformer-sharp: ^4.0.0 => 4.1.0
```

**Comment by LekoArts** [comment]
Hi!

You can follow https://github.com/gatsbyjs/gatsby/pull/33461 for progress on this and/or help getting this merged :)
